packages/flutter/lib/src/widgets/interactive_viewer.dart

```diff
@@ -709,18 +548,150 @@ class _InteractiveViewerState extends State<InteractiveViewer> with TickerProvid
     final Offset focalPointScene = _transformationController!.toScene(
       focalPoint,
     );
-    return matrix
+    final Matrix4 nextMatrix = matrix
       .clone()
       ..translate(focalPointScene.dx, focalPointScene.dy)
       ..rotateZ(-rotation)
       ..translate(-focalPointScene.dx, -focalPointScene.dy);
+
+    return _validateMatrix(nextMatrix, matrix);
+  }
+
+  // If the given matrix is valid, return it. If it is invalid, return the
+  // closest translated matrix that is valid. If none is possible, then return
+  // the given defaultMatrix.
+  //
+  // A valid matrix is defined as a matrix where all sides of the viewport
+  // intersect the boundary. This allows the viewport to partially see beyond
+  // the boundary after rotation, but keeps the entire boundary visible.
+  Matrix4 _validateMatrix(Matrix4 matrix, Matrix4 defaultMatrix) {
```

Comment:
I tried that way first, but I ended up preferring the user experience this way when near the boundary.  If I force the axis aligned bounding box of the viewport to fit inside of the boundary, then I won't be able to see the corners of the child.  Also, I assume that many uses of InteractiveViewer will have a similarly sized boundary and viewport initially, like a photo viewer, and in that case it's nice to be able to rotate at the default size without being stopped by the boundary.

Maybe this is something that could be configured in the long term since the behavior is different for the two approaches.

Reply:
Sorry I probably don't fully understand how this works, but from 

> A valid matrix is defined as a matrix where all sides of the viewport intersect the boundary. 

It sounds like a necessary and sufficient condition for `_boundaryRect` to intersect all 4 edges of `nextViewport` is:
1. `_boundaryRect.top` is in [`nextViewport.smallestY`, `nextViewport.2ndSmallestY`]
2.  `_boundaryRect.bottm` is in [`nextViewport.2ndLargestY`, `nextViewport.largestY`]
3. 4, similarly we can get the horizontal constraints for `_boundaryRect.left` and `_boundaryRect.right`.

`_boundaryRect.bottm - _boundaryRect.top` is a constant. 

So the problem turns into finding the `dx` and `dy` that minimize `dx^2 + dy^2` and satisfy all 4 constaints, or one of them in each direction if they can't be satisfied at the same time, i.e.: 

`_boundaryRect.top + dy` in [`nextViewport.smallestY`, `nextViewport.2ndSmallestY`] and/or [`_boundaryRect.height - nextViewport.largestY`, `_boundaryRect.height - nextViewport.2ndLargestY`]
&&
`_boundaryRect.left + dx` in [`nextViewport.smallestX`, `nextViewport.2ndSmallestX`] and/or [`_boundaryRect.width - nextViewport.largestX`, `_boundaryRect.width - nextViewport.2ndLargestX`]

Reply:
I think you might be on to something, but there is a case where this doesn't work if I understand correctly.

Sorry for the bad drawing but imagine the pen rectangle is the boundary and the pencil rectangle is the viewport:

<img width="300" src="https://user-images.githubusercontent.com/389558/109046042-29deda80-7689-11eb-811f-da96e1b554d5.jpg" />

I would consider this to be invalid because one edge of the viewport does not intersect the boundary.  However, _boundaryRect.top is in [nextViewport.smallestY, nextViewport.2ndSmallestY].

Reply:
There're 4 constraints (2 horizontal and 2 vertical) and to make a viewport valid it has to satisfy all 4. In the drawing the viewport only satisfies 2 (one on each axis): `_boundaryRect.top` is in `[nextViewport.smallestY, nextViewport.2ndSmallestY]` but `_boundaryRect.bottm` is not in `[nextViewport.2ndLargestY, nextViewport.largestY]`

Reply:
ohhh it's sufficient but not necessary. Let me see if it can be improved. I think we can make more use of the fact that the boundary is a `Rect` to simplify the calculation.

Reply:
I think the edges of the viewport must fall in the red area of https://jamboard.google.com/d/1l4Gg2AAs8d_8fGEAK_diDle-9qSQjYYFPkxpzUXAS84/viewer?f=1